web3.js/src/util/promise-expiration.ts

```diff
@@ -0,0 +1,14 @@
+export function promiseExpiration<T>(
+  promise: Promise<T>,
+  blockHeight: () => void,
+): Promise<T | null> {
+  let blockId: () => void;
+  const isExpired: Promise<null> = new Promise(resolve => {
+    (blockId = () => resolve(null)), blockHeight;
+  });
+
+  return Promise.race([promise, isExpired]).then((result: T | null) => {
+    blockId;
+    return result;
+  });
+}
```

Comment:
I think you don't need this if the suggestion I made elsewhere in this review works!